pytorch_lightning/trainer/trainer.py

```diff
@@ -972,7 +972,8 @@ def run_evaluation(self, on_epoch: bool = False) -> _EVALUATE_OUTPUT:
                 dl_outputs = self.track_output_for_epoch_end(dl_outputs, output)
 
             # store batch level output per dataloader
-            self.evaluation_loop.outputs.append(dl_outputs)
+            if self.evaluation_loop.should_track_batch_outputs_for_epoch_end():
+                self.evaluation_loop.outputs.append(dl_outputs)
```

Comment:
@awaelchli @carmocca @tchaton please check me on this. I'm making the assumption that we only need to store the batch-level outputs if we've overridden {validation/test}_epoch_end . this check should save us more memory and make training faster

Reply:
I wonder if you might also want to cache result of `should_track_batch_outputs_for_epoch_end` to avoid doing introspection on every step?